File: eval/fixture-py/invoice.py
"""Invoice math: line totals, bundle unit pricing, and the invoice total.

Applies discount then tax and rounds as it goes, the way a first pass at a
billing service usually grows.
"""

from typing import TypedDict

import money


class Line(TypedDict):
    description: str
    quantity: int
    unit_price: float


def line_total(line: Line) -> float:
    """Total for a single invoice line."""
    return money.round_money(line["unit_price"] * line["quantity"])


def unit_price_from_bundle(bundle_total: float, quantity: int) -> float:
    """Break a bundle price down into a per-unit price."""
    return money.round_money(bundle_total / quantity)


def bundle_line_total(bundle_total: float, quantity: int) -> float:
    """Rebuild a line total from the per-unit bundle price."""
    return money.round_money(unit_price_from_bundle(bundle_total, quantity) * quantity)


def invoice_total(lines: list[Line], discount_pct: float, tax_rate: float) -> float:
    """Sum the lines, take the discount, then add tax."""
    subtotal = 0.0
    for line in lines:
        subtotal += line_total(line)
    discounted = money.round_money(subtotal * (1 - discount_pct / 100))
    taxed = money.round_money(discounted * (1 + tax_rate))
    return taxed

File: eval/fixture-java/src/main/java/com/roastable/billing/Invoice.java
package com.roastable.billing;

import java.util.List;

/**
 * Invoice math: line totals, bundle unit pricing, and the invoice total.
 *
 * <p>Applies discount then tax and rounds as it goes, the way a first pass at a
 * billing service usually grows. Mirrors eval/fixture-py/invoice.py.
 */
public final class Invoice {

    private Invoice() {
    }

    /** A single invoice line. */
    public record Line(String description, int quantity, double unitPrice) {
    }

    /** Total for a single invoice line. */
    public static double lineTotal(Line line) {
        return Money.roundMoney(line.unitPrice() * line.quantity());
    }

    /** Break a bundle price down into a per-unit price. */
    public static double unitPriceFromBundle(double bundleTotal, int quantity) {
        return Money.roundMoney(bundleTotal / quantity);
    }

    /** Rebuild a line total from the per-unit bundle price. */
    public static double bundleLineTotal(double bundleTotal, int quantity) {
        return Money.roundMoney(unitPriceFromBundle(bundleTotal, quantity) * quantity);
    }

    /** Sum the lines, take the discount, then add tax. */
    public static double invoiceTotal(List<Line> lines, double discountPct, double taxRate) {
        double subtotal = 0.0;
        for (Line line : lines) {
            subtotal += lineTotal(line);
        }
        double discounted = Money.roundMoney(subtotal * (1 - discountPct / 100));
        double taxed = Money.roundMoney(discounted * (1 + taxRate));
        return taxed;
    }
}
